libretro cores and libretro kodi support

candrews opened the pull request:
libretro-core.eclass, a number of libretro cores, and then the Kodi addons to use them.

The libretro-core.eclass and cores primarily are from the overlay at https://github.com/stefan-gr/abendbrot - thanks @stefan-gr and @leycec

candrews commented:
@SoapGentoo if you get a chance, can you take a look at this, especially the eclass?

leycec commented:
As the original author of `libretro.eclass` and several of these ebuilds, I wholeheartedly embrace this initiative. All the prior glory goes to @stefan-gr, who single-handedly maintained Gentoo's LibRetro support for nearly a decade on nuthin' but quality German beer, 80's nostalgia, and GitHub stars.

Thanks for fearlessly spearheading their inclusion into Gentoo proper, @candrews. You're awesome.

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
```
eclass doesn't have include guards

candrews replied:
Added:
```
if [[ -z ${_LIBRETRO_CORE_ECLASS} ]]; then
_LIBRETRO_CORE_ECLASS=1
```
to the top and ended the eclass with `fi`

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+
+: ${HOMEPAGE:="https://github.com/${LIBRETRO_REPO_NAME}"}
+
+if [[ ${PV} == "9999" ]]; then
```
`if [[ ${PV} == *9999 ]]; then`
slightly more generalisable (2.1.9999)

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+	: ${EGIT_REPO_URI:="https://github.com/${LIBRETRO_REPO_NAME}.git"}
+	inherit git-r3
+else
+	[ -z "${LIBRETRO_COMMIT_SHA}" ] && die "LIBRETRO_COMMIT_SHA must be set before inherit."
```
please no single brackets, we don't need to support POSIX here

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+	6)
+		EXPORT_FUNCTIONS src_unpack src_prepare src_compile src_install
+		;;
+        *) die "EAPI=${EAPI} is not supported" ;;
```
use the same indentation as the `6)` block

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+# This function retrieves the remote Libretro core info files.
+libretro-core_src_unpack() {
+	# If this is a live ebuild, retrieve this core's remote repository.
+	if [[ ${PV} == "9999" ]]; then
```
ditto

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+		# * Where SHARED is used to provide shared linking
+		#   flags ensure final link command includes LDFLAGS
+		#   and LIBS
+		# * Always use $(CFLAGS) when calling $(CC)
```
can't we work with upstream to use a sane buildsystem instead of awful homebrew makefiles?

candrews replied:
Maybe? I'm not quite sure I'm ready to boil that ocean right now though... the problem exists in a dozen repositories, not all of which are under the control of the same organization.

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+libretro-core_src_compile() {
+	emake CC=$(tc-getCC) CXX=$(tc-getCXX) \
+		$(usex debug "DEBUG=1" "") "${myemakeargs[@]}" \
+		$([ -f makefile.libretro ] && echo '-f makefile.libretro') \
```
single brackets

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+	emake CC=$(tc-getCC) CXX=$(tc-getCXX) \
+		$(usex debug "DEBUG=1" "") "${myemakeargs[@]}" \
+		$([ -f makefile.libretro ] && echo '-f makefile.libretro') \
+		$([ -f Makefile.libretro ] && echo '-f Makefile.libretro')
```
ditto

candrews replied:
:+1: done!

SoapGentoo commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,162 @@
+# This function installs the shared library for this Libretro core.
+libretro-core_src_install() {
+	# Absolute path of the directory containing Libretro shared libraries.
+	LIBRETRO_LIB_DIR="/usr/$(get_libdir)/libretro"
```
what about `${EPREFIX}`?

candrews replied:
Changed to `LIBRETRO_LIB_DIR="${EPREFIX}/usr/$(get_libdir)/libretro"`

floppym commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,168 @@
+# This function installs the shared library for this Libretro core.
+libretro-core_src_install() {
+	# Absolute path of the directory containing Libretro shared libraries.
+	LIBRETRO_LIB_DIR="${EPREFIX}/usr/$(get_libdir)/libretro"
```
This was actually correct without `EPREFIX`; `insinto` automatically adds EPREFIX to the path `EPREFIX`. Please revert the change.

candrews replied:
Done.

The documentation at https://devmanual.gentoo.org/function-reference/install-functions/index.html should probably be updated. I, at least, get tripped up frequently on the use of EPREFIX as related to the install functions.

candrews commented:
@mgorny I've addressed the feedback you provided on the gentoo-dev list at https://archives.gentoo.org/gentoo-dev/message/0bdc848b97ede9bd294be8b2fc8f6ae0 - can you please take a look again and see if there are further improvements you can/want to suggest?

mgorny commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+	local LIBRETRO_LIB_DIR="/usr/$(get_libdir)/libretro"
+	# If this core's shared library exists, install that.
+	if [[ -f "${LIBRETRO_CORE_LIB_FILE}" ]]; then
+		insinto "${LIBRETRO_LIB_DIR}"
```
`exeinto` maybe?

candrews replied:
agreed

mgorny commented on games-emulation/libretro-nestopia/libretro-nestopia-0.0.1_pre20180723.ebuild:
```diff
@@ -0,0 +1,15 @@
+LIBRETRO_COMMIT_SHA="9c335527c5dfa11dfa29bc22ee4ba6efa23c8890"
+KEYWORDS="~amd64 ~x86"
+inherit libretro-core
+S="${S}/libretro"
```
Don't use S om S, this will go crazy if ebuild is sourced multiple times.

candrews replied:
Are you saying that I shouldn't use S within S? If that's true, that `S="${WORKDIR}/${P}/libretro"` would be okay, so I've done that.

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+libretro-core_src_prepare() {
+	ebegin "Attempting to hack Makefiles to use custom cflags"
+	# Populate COMMIT for GIT_VERSION
+	CUSTOM_LIBRETRO_COMMIT_SHA="\" ${LIBRETRO_COMMIT_SHA:0:7}\""
```
Shouldn't be, thanks for the review.

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+		#   and LIBS
+		# * Always use $(CFLAGS) when calling $(CC)
+		# * Add short-rev to Makefile
+		[[ -e "${makefile}" ]] && sed \
```
You're right!

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+#
+# This function prepares the source by making custom modifications.
+libretro-core_src_prepare() {
+	ebegin "Attempting to hack Makefiles to use custom cflags"
```
I would be fast, yes. I don't really have any strong feelings on the `ebegin`/`eend`, so I've removed them.

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+	${shopt_saved}
+	eend
+	export OPTFLAGS="${CFLAGS}"
+	default_src_prepare
```
Great point!

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+# src_compile() {
+#       local myemakeargs=(
+#		$(usex neon "HAVE_NEON=1" "")
+#	)
```
:+1:

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+#       local myemakeargs=(
+#		$(usex neon "HAVE_NEON=1" "")
+#	)
+#       libretro-core_src_configure
```
:+1:

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+	# to enable additional debug features.
+	emake CC=$(tc-getCC) CXX=$(tc-getCXX) \
+		$(usex debug "DEBUG=1" "") "${myemakeargs[@]}" \
+		$([[ -f makefile.libretro ]] && echo '-f makefile.libretro') \
```
At most one should exist; I haven't been able to come up with a decent one-liner that does that. I tried:
```
        emake CC=$(tc-getCC) CXX=$(tc-getCXX) \
                $(usex debug "DEBUG=1" "") "${myemakeargs[@]}" \
                -f [Mm]akefile.libretro
```
But that doesn't work if `[Mm]akefile.libretro` doesn't exist. I also tried:
```
$([[ -f [Mm]akefile.libretro ]] && echo '-f [Mm]akefile.libretro')
```
but that doesn't work because there is no file name expansion inside `[[ ]]``

So, I think the two lines I originally had are the most readable solution.

mgorny replied:
I suppose you could do

    $([[ -f Makefile.libretro || -f makefile.libretro ]] && echo -f [Mm]akefile.libretro)

but I agree that's probably less readable.

candrews commented on eclass/libretro-core.eclass:
```diff
@@ -0,0 +1,197 @@
+# This function installs the shared library for this Libretro core.
+libretro-core_src_install() {
+	# Absolute path of the directory containing Libretro shared libraries.
+	local LIBRETRO_LIB_DIR="/usr/$(get_libdir)/libretro"
```
done

candrews commented:
@mgorny I've made all the changes you suggested, including adjusting the commit messages.

Thank you for the review!

jason-oliveira commented:
hi. casting a little necromancy on this.
